### PR TITLE
Add second transactions api provider

### LINKED ISSUE #1: Add another block explorer API
It currently uses only etherscan API, it would be ideal to get the transaction from more that one provider and compare the tx list received in the same from both services.

### PR DESCRIPTION
**Pull request by ethereal-nfts**
WIP. Currently adds Blockscout.
Fixes #1

### CODE REVIEW

**Comment by AugustoL** [comment]
@ethereal-nfts great! let me know here when I can review!

**Comment by ethereal-nfts** [comment]
Very strange, successfully pulling tx and from api but internal tx are different. Standard tx are identical.
3859 results for 
[https://blockscout.com/eth/mainnet/api?module=account&action=txlistinternal&address=0x9f828AC3baA9003e8a4e0b24bcaE7b027B6740b0&startblock=7850000&endblock=10162933&sort=asc](https://blockscout.com/eth/mainnet/api?module=account&action=txlistinternal&address=0x9f828AC3baA9003e8a4e0b24bcaE7b027B6740b0&startblock=7850000&endblock=10162933&sort=asc )
0 results for
[https://api.etherscan.io/api?module=account&action=txlistinternal&address=0x9f828AC3baA9003e8a4e0b24bcaE7b027B6740b0&startblock=7850000&endblock=10162933&sort=asc](https://api.etherscan.io/api?module=account&action=txlistinternal&address=0x9f828AC3baA9003e8a4e0b24bcaE7b027B6740b0&startblock=7850000&endblock=10162933&sort=asc)

Will commit code as it appears to be func

**Comment by ethereal-nfts** [comment]
last commit adds cross checking for standard tx. All pass up to genesisProtocol which has 3462 from etherscan and 3453 from blockscout.

**Comment by ethereal-nfts** [comment]
@AugustoL This PR is ready to view. Important to note that BlockScout is discovering far more (x000's) internal tx than etherscan.

**Comment by AugustoL** [comment]
> @AugustoL This PR is ready to review. Important to note that BlockScout is discovering far more (x000's) internal tx than etherscan.

Good catch, for now we will filter and only use the matching txs.

File: scripts/get-transactions.js [hunk 1]
```diff
@@ -1,285 +1,376 @@
...
-        return contractEvent.returnValues._avatar && (contractEvent.returnValues._avatar != dxAvatar.address)
-      });
-    }
-  }
-  
-  fs.writeFileSync('DXdaoSnapshot.json', JSON.stringify(DXdaoSnapshot, null, 2), {encoding:'utf8',flag:'w'});
-} 
-
-Promise.all([main()]).then(process.exit);
+const fs = require('fs');
```
Comment (AugustoL) [inline scripts/get-transactions.js:1]: Weird that I see the whole file rewritten..is difficult to review it that way.

File: scripts/get-transactions.js [hunk 2]
```diff
@@ -1,285 +1,376 @@
...
+    }
+  }
+
+  function etherscanGetHashFromApiServiceTx(tx){
+    return tx.hash
+  }
+
+  function etherscanTxsUriBuilder(_address, _fromBlock, _toBlock){
+    return 'https://api.etherscan.io/api?module=account&action=txlist&address='
+            +_address
```
Comment (AugustoL) [inline scripts/get-transactions.js:210]: The indentation here is wrong, we use two spaces indentation in the script.

File: scripts/get-transactions.js [hunk 3]
```diff
@@ -1,285 +1,376 @@
...
+    return 'https://api.etherscan.io/api?module=account&action=txlistinternal&address='
+            +_address
+            +'&startblock='+_fromBlock
+            +'&endblock='+_toBlock
+            +'&sort=asc&apikey='
+            +EtherscanAPIToken
+  }
+
+  function blockscoutGetHashFromApiServiceTx(tx){
+    if(typeof(tx.hash)!=='undefined') return tx.hash
```
Comment (AugustoL) [inline scripts/get-transactions.js:227]: Missing space between teh if and ( and return goes after the break line, no need to do one lines ifs :)

File: scripts/get-transactions.js [hunk 4]
```diff
@@ -1,285 +1,376 @@
...
+    		});
+    	});
+  		return JSON.parse(await http_promise).result;
+  	}
+  	catch(error) {
+  		console.error(error);
+  	}
+  }
+
+  async function asyncForEach(array, callback) {
```
Comment (AugustoL) [inline scripts/get-transactions.js:177]: This function is not used, it can be removed.

File: scripts/get-transactions.js [hunk 5]
```diff
@@ -1,285 +1,376 @@
...
+  }
+
+  function blockscoutGetHashFromApiServiceTx(tx){
+    if(typeof(tx.hash)!=='undefined') return tx.hash
+    if(typeof(tx.transactionHash)!=='undefined') return tx.transactionHash
+    console.log('Failed tx: ',tx)
+    throw 'Neither tx.hash nor tx.transactionHash found in blockscout tx.'
+  }
+
+  function blockscoutTxsUriBuilder(_address,_fromBlock,_toBlock){
```
Comment (AugustoL) [inline scripts/get-transactions.js:233]: MIssing spaces after , and between ){
```
  function blockscoutTxsUriBuilder(_address, _fromBlock, _toBlock) {
```

**Comment by AugustoL** [comment]
@ethereal-nfts did you had any issue with the API rate limits when you run it in -reset mode? did you tested it? :eyes:

File: scripts/get-transactions.js (hunk 1 above)
Comment (ethereal-nfts) [reply scripts/get-transactions.js:1]: I'm not sure why this is... very strange.

**Comment by ethereal-nfts** [comment]
> 
> 
> @ethereal-nfts did you had any issue with the API rate limits when you run it in -reset mode? did you tested it? 👀

Yes, it's properly generating the new DXdaoSnapsjot.json (3x bigger) at 20 req/s. Specific command I'm using is `node scripts/get-transactions.js -network mainnet -reset`. It does seem Blockscout occasionally returns an http error when run for the first time instead of the json, but not due to rate limiting. I'll add a check for this exception.

**Comment by ethereal-nfts** [comment]
> 
> 
> KEY_BLOCKCYPHER is missing in .env.example file
Currently using BlockScout instead of Blockcypher, so key not needed.

**Comment by ethereal-nfts** [comment]
Reopen (accidentally clicked wrong button)

**Comment by ethereal-nfts** [comment]
Commit [2a58dcf](https://github.com/AugustoL/dxdao-snapshot/pull/8/commits/2a58dcf9cd68f784b1903368edbd82381f89605f) fixes the temporary server side timeout issues for large json requests from blockscout.

**Pull request merged**